File: 算法训练营/03-数组链表跳表/070/缓存法.go
/*
 * @lc app=leetcode.cn id=70 lang=golang
 *
 * [70] 爬楼梯
 */

// @lc code=start
func climbStairs(n int) int {
	records := make(map[int]int)
	return helper(n, records)
}

func helper(n int, records map[int]int) int {
	if n < 0 {
		return 0
	} else if n <= 3 {
		return n
	}

	ret, ok := records[n]
	if ok {
		return ret
	}
	ret = helper(n-1, records) + helper(n-2, records)
	records[n] = ret
	return ret
}
// @lc code=end

File: 算法训练营/03-数组链表跳表/070/缓存法.py
#
# @lc app=leetcode.cn id=70 lang=python3
#
# [70] 爬楼梯
#

# @lc code=start
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [1, 2]
        for i in range(2, n):
            dp.append(dp[i-1] + dp[i-2])
        return dp[n-1]
# @lc code=end